gcode-text.1
.TH GCODE-TEXT 1
.SH NAME
gcode-text \- render text as gcode
.SH SYNOPSIS
gcode-text [options] [input files] ...
.SH DESCRIPTION
.PP
\fIGcode-text\fP renders stroked text to gcode. Each line of text is
rendered to separate rectangle, where rectangles may be defined
using a template or parametrically on the command line.
.PP
The source text may come from text files or from command line
arguments.
.PP
The provided font, Twin Sans, uses glyphs originally derived from the
Hershey fonts.
.SH USAGE
.PP
An un-flagged argument is treated as an input file.
.PP
Options are as follows:
.TP
.B "--help,--usage"
Print a help/usage message and exit.
.TP
.BI "-V,--version"
Prints the version number and immediately exits.
.TP
.BI "-i,--inch"
Use inch units in the generated file. This is the default.
.TP
.BI "-m,--mm"
Use millimeter units in the generated file.
.TP
.BI "-r,--rect"
Draw bounding rectangles
.TP
.BI "-O,--oblique"
Draw the glyphs using a sheer transform
.TP
.BI "--tesselate"
Tesselate splines even when the output device supports them natively.
.TP
.BI "--sheer " sheer
Set the sheer amount for oblique glyphs.
.TP
.BI "-f,--flatness " flatness
Set the spline decomposition tolerance in output units. The default is
0.001.
.TP
.BI "--font " SVG-font-file
Specify the filename of the font to draw with. Default is TwinSans.svg.
.TP
.BI "--feed " feed
Set the feed rate. Note that this value depends on the units in use.
.TP
.BI "--speed " speed
Set the spindle speed. Nominally in RPM, check your device documentation
for non-rotating tools to understand what it does.
.TP
.BI "-t,--template " template
Specifies a template file. Template files are in JSON format and
consist of an array of rectangles, each rectangle is an array of
four values -- [x, y, width, height].
.TP
.BI "-d,--device " device-spec
Specifies a device file. This is a JSON file containing
definitions for drawing to the target. See the DEVICE SPEC section
below.
.TP
.BI "-S,--settings " setting,...
Specifies a comma-separated list of device-specific settings. See
the "settings" and "setting-values" items in the DEVICE SPEC section below.
.TP
.BI "-o,--output " output
Specifies the output file name.
.TP
.BI "-b,--border " border
Defines the minimum space between text and each bounding
box. The default is 0.1.
.TP
.BI "-x,--start-x " start-x
Defines the starting X location for parametrically generated boxes.
.TP
.BI "-y,--start-y " start-y
Defines the starting Y location for parametrically generated boxes.
.TP
.BI "-w,--width " width
Defines the width for parametrically generated boxes.
.TP
.BI "-h,--height " height
Defines the height for parametrically generated boxes.
.TP
.BI "-X,--delta-x " delta-x
Defines the X offset between parametrically generated boxes.
.TP
.BI "-Y,--delta-y " delta-y
Defines the Y offset between parametrically generated boxes.
.TP
.BI "--final-x " final-x
Defines the X coordinate to move after the drawing is finished.
.TP
.BI "--final-y " final-y
Defines the Y coordinate to move after the drawing is finished.
.TP
.HI "--down " down-z
Defines the Z position when drawing strokes, for devices with a Z axis.
.TP
.HI "--up " up-z
Defines the Z position when moving between strokes, for devices with a Z axis.
.TP
.BI "-c,--columns " columns
Defines the number of columns of parametrically generated boxes.
.TP
.BI "-v,--value " value
Sets an initial numeric value for generated text.
.TP
.BI "-n,--number " number
Specifies the number of generated numeric text values
.TP
.BI "-T,--text " string
Specifies a single line of text
.TP
.BI "-a,--align " {left,right,center}
Specifies the alignment of text within the box.
.TP
.BI "--font-metrics
Use font metrics (font height and string width) instead of ink bounds
for text layout.
.TP
.BI "-C,--config-dir " directory
Specifies an additional path to device configuration files. This
argument may be repeated.
.SH EXAMPLES
.PP
Draw "hello world" into a 4"x1" box at 1"x1"
.nf
$ gcode-text -o output.gcode -T "hello world" -w 4 -h 1 -x 1 -y 1
.fi
.PP
Draw numbers starting at 1 using the sample template
.nf
$ gcode-text -o output.gcode -v 1 -t sample-template.json
.fi
.PP
Draw 10 numbers from 10232 to 10241 in a 2x5 grid of 0.25"x0.1" boxes
spaced 4"x1.5" apart and offset by 1.5"x1" from the origin with a
border of 0.01" around the text:
.nf
$ gcode-text -o output.gcode -v 10232 -n 10 -c 2 -x 1.5 -y 1 -w .25 -h .1 -X 4 -Y 1.5 -b 0.01
.fi
.SH DEVICE SPEC
.PP
This file contains device-specific output customization, providing
control over the strings used to drive the target. This JSON file
should contain a single dictionary with keys mapping to string
values. Each value can contain embedded newlines as necessary;
gcode-text doesn't emit any newlines itself. If not provided in the
configuration file (or if no configuration file is provided), the
default values shown below will be used instead.
.TP
.BI "start " "G90\\\\\\\\nG17\\\\\\\\n"
Specifies a string emitted before any drawing is done to set up the
target.
.TP
.BI "settings " "\"\""
Specifies a C-style printf format for outputting device-specific
settings. If the string is empty, this element is skipped.
.TP
.BI "setting-values " "\"\""
Specifies an array of strings, one for each %s value present in the
"settings" value. These can be replaced on the command line using
the --settings option.
.TP
.BI "inch " "G20\\\\\\\\n"
Specifies the string to place the device in "inch" mode.
.TP
.BI "mm " "G21\\\\\\\\n"
Specifies the string to place the device in "millimeter" mode.
.TP
.BI "move " "G00 X%f Y%f\\\\\\\\n"
Specifies a C-style printf format for repositioning the output without
drawing. This format should contain two %f instances, the first is
used for the X coordinate, the second for Y.
.TP
.BI "draw " "G01 X%f Y%f F%f S%f\\\\\\\\n"
Specifies a C-style printf format for drawing a straight line. This
format should contain two to four %f instances, the first is used for the X
coordinate, the second for Y, followed by an (optional) feed rate and
(optional) spindle speed.
.TP
.HI "z-move " "G01 Z%f F%f S%f\\\\\\\\n"
Specifies a C-style printf format for moving the tool in the Z
axis. This
format should contain one to three %f instances, followed by an
(optional) feed rate and (optional) spindle speed.
.TP
.BI "feed " "true or false"
Indicates whether the draw element includes a %f required for
the feed rate.
.TP
.BI "speed " "true or false"
Indicates whether the draw element includes a %f required for
the spindle speed.
.TP
.BI "z-axis" "true or false"
Indicates whether the machine requires Z-axis motions when
transitioning between draw and move operations. This is 'false' by default'.
.TP
.BI "y-invert " "true or false"
Indicates whether y values indicate upwards (true) or downwards
(false) motion.
.TP
.BI "stop " "M30\\\\\\\\n"
Specifies the string emitted after all drawing is done.
.PP
Gcode-text first emits the "start" string, then one of "inch" or
"mm". Then, drawing is done by emitting a sequence of "move" and
"draw" operations. Finally, drawing is finished with the "stop" string
and the output file is closed.
.PP
Here's what the default configuration looks like in JSON form:
.nf
{
    "start" : "G90\nG17\\n",
    "inch" : "G20\\n",
    "mm" : "G10\\n",
    "move" : "G00 X%f Y%f\\n",
    "draw" : "G01 X%f Y%f F%f S%f\\n",
    "feed" : true,
    "speed" : true,
    "stop" : "M30\\n"
}
.fi
.SH AUTHOR
\fIGcode-text\fP is the work of Keith Packard <keithp@keithp.com>.
.\"
.PP
Gcode-text is
.br
Copyright 2023 Keith Packard.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.
.PP
You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,


debian/changelog
gcode-text (0.8-1) unstable; urgency=medium

  * Add --x-off and --y-off to gcode-svg
  * Match colors by value instead of by string in gcode-svg

 -- Keith Packard <keithp@keithp.com>  Sun, 17 May 2026 21:54:27 -0700

gcode-text (0.7-1) unstable; urgency=medium

  * Add Z motion control
  * Add scan conversion for filled areas

 -- Keith Packard <keithp@keithp.com>  Sun, 17 May 2026 14:09:45 -0700

gcode-text (0.6-2) unstable; urgency=medium

  * Make python dependencies explicit

 -- Keith Packard <keithp@keithp.com>  Sun, 31 Dec 2023 17:19:32 -0800

gcode-text (0.6-1) unstable; urgency=medium

  * Add gcode-svg
  * Add script font
  * Fix text centering in y-invert mode.
  * Add parametric template support
  * Lots of font updates
  * Switch fonts to svg format

 -- Keith Packard <keithp@keithp.com>  Sun, 31 Dec 2023 14:22:55 -0800

gcode-text (0.5-1) unstable; urgency=medium

  * Switch to python
  * Change template to JSON format
  * Add --sheer and --tesselate options

 -- Keith Packard <keithp@keithp.com>  Sun, 03 Dec 2023 17:26:02 -0800

gcode-text (0.4-1) unstable; urgency=medium

  * Add spline output support
  * Add "settings" to device configurations
  * Add --version and --oblique options

 -- Keith Packard <keithp@keithp.com>  Fri, 01 Dec 2023 13:36:47 -0800

gcode-text (0.3-1) unstable; urgency=medium

  * Add SVG output support

 -- Keith Packard <keithp@keithp.com>  Thu, 30 Nov 2023 16:47:13 -0800

gcode-text (0.2-1) unstable; urgency=medium

  * Add device config file support.

 -- Keith Packard <keithp@keithp.com>  Wed, 29 Nov 2023 17:36:28 -0800

gcode-text (0.1-1) unstable; urgency=medium

  * Initial release. (Closes: #123456)

 -- Keith Packard <keithp@keithp.com>  Wed, 29 Nov 2023 12:04:50 -0800


gcode-text.py
#!/usr/bin/env python3
#
# Copyright © 2023 Keith Packard <keithp@keithp.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
#

from __future__ import annotations
import math
import json
import sys
import argparse
import csv
import os
import numbers
from typing import Any
from io import StringIO
from lxml import etree  # type: ignore
from svg.path import parse_path  # type: ignore
from svg.path import Path, Move, Line, Arc, CubicBezier, QuadraticBezier, Close  # type: ignore

if not os.getenv("GCODE_SKIP_PATH"):
    sys.path = ["@SHARE_DIR@"] + sys.path

from gcode_draw import *


class TextValues(Values):

    def __init__(self):
        super().__init__()
        self.font = "TwinSans.svg"
        self.oblique = False
        self.border = 0
        self.start_x = None
        self.start_y = None
        self.width = 4
        self.height = 1
        self.delta_x = 4
        self.delta_y = 1
        self.up = 0
        self.down = 0.125
        self.columns = 1
        self.value = None
        self.number = 1
        self.text = None
        self.align = "center"
        self.font_metrics = False
        self.rects = None
        self.file = None


def Args():
    parser = argparse.ArgumentParser(add_help=False, description="Render stroked text")
    Device.args(parser)
    parser.add_argument(
        "-r",
        "--rect",
        action="store_true",
        help="Draw bounding rectangles",
        default=None,
    )
    parser.add_argument(
        "-O",
        "--oblique",
        action="store_true",
        help="Draw the glyphs using a sheer transform",
        default=None,
    )
    parser.add_argument(
        "--sheer", action="store", type=float, help="Oblique sheer amount"
    )
    parser.add_argument(
        "--font", action="store", type=str, help="SVG font file name", default=None
    )
    parser.add_argument(
        "-t", "--template", action="store", help="Template file name", default=None
    )
    parser.add_argument(
        "-b", "--border", action="store", type=float, help="Border width"
    )
    parser.add_argument(
        "-x", "--start-x", action="store", type=float, help="Starting X for boxes"
    )
    parser.add_argument(
        "-y", "--start-y", action="store", type=float, help="Starting Y for boxes"
    )
    parser.add_argument("-w", "--width", action="store", type=float, help="Box width")
    parser.add_argument("-h", "--height", action="store", type=float, help="Box height")
    parser.add_argument(
        "-X", "--delta-x", action="store", type=float, help="X offset between boxes"
    )
    parser.add_argument(
        "-Y", "--delta-y", action="store", type=float, help="Y offset between boxes"
    )
    parser.add_argument(
        "--final-x", action="store", type=float, help="Final X position"
    )
    parser.add_argument(
        "--final-y", action="store", type=float, help="Final Y position"
    )
    parser.add_argument(
        "-c", "--columns", action="store", type=int, help="Number of columns of boxes"
    )
    parser.add_argument(
        "-v", "--value", action="store", type=float, help="Initial text numeric value"
    )
    parser.add_argument(
        "-n", "--number", action="store", type=float, help="Number of numeric values"
    )
    parser.add_argument("-T", "--text", action="store", help="Text string")
    parser.add_argument("--up", action="store", type=float, help="Pen up Z position")
    parser.add_argument(
        "--down", action="store", type=float, help="Pen down Z position"
    )
    parser.add_argument(
        "-a",
        "--align",
        action="store",
        type=str,
        choices=["left", "right", "center"],
        default=None,
    )
    parser.add_argument(
        "--font-metrics",
        action="store_true",
        help="Use font metrics for strings instead of glyph metrics",
        default=None,
    )
    parser.add_argument(
        "--dump-stf", action="store", help="Dump font in STF format", default=None
    )
    parser.add_argument("file", nargs="*", help="Text source files")
    args = parser.parse_args()

    if args.help:
        parser.print_help()
        sys.exit(0)

    if args.version:
        print("%s" % "@VERSION@")
        sys.exit(0)

    return args


def finite_rects(values):
    return values.rects is not None


def load_template(template_file, values):

    with values.config_open(template_file) as file:
        template = json.load(file)
    if isinstance(template, list):
        value.rects = template
    elif isinstance(template, dict):
        values.handle_dict(template)

    if values.rects is None:
        return

    if not isinstance(values.rects, list):
        print("template rects is not an array", file=sys.stderr)
        raise TypeError
    for e in tuple(values.rects):
        if not isinstance(e, list):
            print("rects element %s is not an array" % (e,), file=sys.stderr)
            raise TypeError
        if len(e) != 4:
            print(
                "rects element %s does not contain four values" % (e,), file=sys.stderr
            )
            raise TypeError
        for v in tuple(e):
            if not isinstance(v, numbers.Number):
                print("rects value %r is not a number" % (v,), file=sys.stderr)
                raise TypeError


def get_rect(values):
    if values.rects is not None:
        for r in values.rects:
            yield Rect(Point(r[0], r[1]), Point(r[0] + r[2], r[1] + r[3]))
    else:
        y = values.start_y
        while True:
            x = values.start_x
            for c in range(values.columns):
                yield Rect(Point(x, y), Point(x + values.width, y + values.height))
                x += values.delta_x
            y += values.delta_y


def get_line(values):
    if values.value != None:
        v = values.value
        n = values.number
        while finite_rects(values) or n > 0:
            yield "%d" % v
            n -= 1
            v += 1
    if values.text != None:
        for l in values.text.splitlines():
            yield l
    for name in values.file:
        with open(name, "r", encoding="utf-8", errors="ignore") as f:
            for l in f.readlines():
                yield l.strip()


def text_path(gcode: GCode, m: Matrix, s: str):
    draw = MatrixDraw(gcode.get_draw(), m)
    gcode.font.text_path(s, draw)


def text_into_rect(gcode: GCode, r: Rect, s: str, values: TextValues):
    if gcode.values.rect:
        gcode.rect(r)

    rect_width = r.bottom_right.x - r.top_left.x - values.border * 2
    rect_height = r.bottom_right.y - r.top_left.y - values.border * 2

    if rect_width < 0:
        print("border %f too wide for rectangle %s" % (values.border, r))
        return
    if rect_height < 0:
        print("border %f too tall for rectangle %s" % (values.border, r))
        return

    metrics = gcode.font.text_metrics(s)

    if values.font_metrics:
        ascent = gcode.font.ascent
        descent = gcode.font.descent
        text_x: float = 0
        text_width = metrics.width
    else:
        ascent = metrics.ascent
        descent = metrics.descent
        text_x = metrics.left_side_bearing
        text_width = metrics.right_side_bearing - metrics.left_side_bearing

    text_height = ascent + descent

    if text_width == 0 or text_height == 0:
        print("Text is empty")
        return

    if values.oblique:
        text_width += text_height * values.sheer

    if text_width / text_height > rect_width / rect_height:
        scale = rect_width / text_width
    else:
        scale = rect_height / text_height

    text_off_y = (rect_height - text_height * scale) / 2

    if values.align == "left":
        text_off_x: float = 0
    elif values.align == "center":
        text_off_x = (rect_width - text_width * scale) / 2
    else:
        text_off_x = text_width

    metrics_x_adjust = text_x * scale

    text_off_x = text_off_x - metrics_x_adjust

    text_x = text_off_x + r.top_left.x + values.border
    text_y = text_off_y + r.top_left.y + values.border
    text_x_span = text_width * scale
    text_y_span = text_height * scale

    matrix = Matrix()
    matrix = matrix.translate(
        text_off_x + r.top_left.x + values.border,
        text_off_y + r.top_left.y + values.border,
    )
    if values.oblique:
        matrix = matrix.sheer(-values.sheer, 0)

    matrix = matrix.scale(scale, scale)
    if gcode.device.y_invert:
        matrix = matrix.translate(0, descent)
        matrix = matrix.scale(1, -1)
    else:
        matrix = matrix.translate(0, ascent)

    text_path(gcode, matrix, s)


def main():
    values = TextValues()
    args = Args()

    if args.config_dir:
        values.config_dir = args.config_dir + values.config_dir

    if args.template:
        load_template(args.template, values)

    values.handle_args(args)

    device = Device(values)

    font = Font.svg_font(values.font, values)

    if args.dump_stf:
        with open(args.dump_stf, "w") as file:
            font.dump_stf(file)
            print("", file=file)
        sys.exit(0)

    rect_gen = get_rect(values)
    line_gen = get_line(values)

    output = sys.stdout
    if args.output != "-":
        output = open(args.output, "w")

    gcode = GCode(output, device, values, font)
    gcode.start()

    while True:
        try:
            rect = next(rect_gen)
            line = next(line_gen)
            text_into_rect(gcode, rect, line, values)
        except StopIteration:
            break

    gcode.stop()


main()
